libraries/hotstuff/chain_pacemaker.cpp

```diff
@@ -112,9 +112,31 @@ namespace eosio { namespace hotstuff {
       return _chain->is_builtin_activated( builtin_protocol_feature_t::instant_finality );
    }
 
-   void chain_pacemaker::get_state( finalizer_state& fs ) const {
-      if (enabled())
-         _qc_chain.get_state( fs ); // get_state() takes scare of finer-grained synchronization internally
+   void chain_pacemaker::get_state(finalizer_state& fs) const {
+      if (! enabled())
+         return;
+
+      // lock-free state version check
+      uint64_t current_state_version = _qc_chain.get_state_version();
+      if (_state_cache_version != current_state_version) {
+         finalizer_state current_state;
+         {
+            csc prof("stat");
+            std::lock_guard g( _hotstuff_global_mutex ); // lock IF engine to read state
+            prof.core_in();
+            _qc_chain.get_state(current_state);
+            current_state_version = _qc_chain.get_state_version(); // get potentially fresher version
+            prof.core_out();
+         }
+         {
```

Comment:
```suggestion
            current_state_version = _qc_chain.get_state_version(); // get potentially fresher version
            if (_state_cache_version != current_state_version) 
               _qc_chain.get_state(current_state);
            prof.core_out();
         }
         if (_state_cache_version != current_state_version) {
```

Reason for this change is that multiple http threads could be calling this at the same time, and end up waiting on the lock at line 125. After the first thread has updated the cache, there is no need for the other threads to do the same thing.